test/hotspot/jtreg/compiler/lib/compile_framework/CompileFramework.java

```diff
@@ -0,0 +1,303 @@
+        try {
+            Process process = builder.start();
+            boolean exited = process.waitFor(COMPILE_TIMEOUT, TimeUnit.SECONDS);
+            if (!exited) {
+                process.destroyForcibly();
+                System.out.println("Timeout: compile command: " + String.join(" ", command));
+                throw new InternalCompileFrameworkException("Process timeout: compilation took too long.");
+            }
+            output = new String(process.getInputStream().readAllBytes(), StandardCharsets.UTF_8);
+            exitCode = process.exitValue();
+        } catch (IOException e) {
+            throw new InternalCompileFrameworkException("IOException during compilation", e);
+        } catch (InterruptedException e) {
+            throw new CompileFrameworkException("InterruptedException during compilation", e);
+        }
+
+        if (exitCode != 0 || !output.equals("")) {
+            System.out.println("Compilation failed.");
+            System.out.println("Exit code: " + exitCode);
+            System.out.println("Output: '" + output + "'");
+            throw new CompileFrameworkException("Compilation failed.");
+        }
+    }
+
+    private void setUpClassLoader() {
+        ClassLoader sysLoader = ClassLoader.getSystemClassLoader();
+
+        try {
+            // Classpath for all included classes (e.g. IR Framework).
+            // Get all class paths, convert to urls.
+            List<URL> urls = new ArrayList<URL>();
+            for (String path : getClassPaths()) {
+                urls.add(new File(path).toURI().toURL());
+            }
+            // And add in the compiled classes from this instance of CompileFramework.
+            urls.add(new File(classesDir.toString()).toURI().toURL());
+            classLoader = URLClassLoader.newInstance(urls.toArray(URL[]::new), sysLoader);
+        } catch (IOException e) {
+            throw new CompileFrameworkException("IOException while creating ClassLoader", e);
+        }
+    }
+
+    public Class getClass(String name) {
+        try {
+            return Class.forName(name, true, classLoader);
+        } catch (ClassNotFoundException e) {
+            throw new CompileFrameworkException("Class not found:", e);
+        }
+    }
+
+    public Object invoke(String className, String methodName, Object[] args) {
+        Class c = getClass(className);
+
+        Method[] methods = c.getDeclaredMethods();
+
+        Method method = null;
+
+        for (Method m : methods) {
+            if (m.getName().equals(methodName)) {
+                if (method != null) {
+                  throw new CompileFrameworkException("Method name \"" + methodName + "\" not unique in class \n" + className + "\".");
+                }
+                method = m;
+            }
+        }
+
+        if (method == null) {
+            throw new CompileFrameworkException("Method \"" + methodName + "\" not found in class \n" + className + "\".");
+        }
```

Comment:
Could be extracted to a method which returns a non-null method object (i.e. `findMethod()`) or something like that.